### ISSUE TITLE
Wire v2 Working Document

### ISSUE THREAD

**Issue by JakeWharton** [square/wire#96]
Since multiple people are expressing interest in pushing this along I thought it best to write down what we want v2 to be. This will ensure everyone is on the same page and allow parallel work to be done. This is meant to be a living document that will be edited as things change.

### Summary:
 * Simplify use in a toolchain.
 * Leverage parsing & pruning to generate things other than Wire-specific Java code.
 * Alter the behavior of the Wire-specific generated Java code.
 * More advanced CLI
 * Other features

#### Simplify use in a toolchain.

The Wire compiler has no API which means you are forced to interact with by calling the `main()` method with CLI arguments.

For example,
```groovy
task generateProtos << {
  def protoRoots = project.file("proto_roots.txt").readLines()
  def protoFiles = project.file("proto_files.txt").readLines()
  def compiler = new WireCompiler(protoDir.toString(), protoFiles, protoRoots, 'src/main/java', null, false)
  compiler.compile()
}
```

This is unsatisfactory. The tests for wire are also forced into doing this.

We should provide an API which lets you very easily specify all of the required data for compilation.

Random example that's not a suggestion of what the actual API should be:
```java
new WireCompiler()
    .addDirectory(new File("/foo/bar/protos"))
    .addProtoRoot("example.thing.Foo")
    .addProtoRoot("example.thing.Bar")
    .build();
```

*(This is done on [version-two branch][branch])*

#### Leverage parsing & pruning to generate things other than Wire-specific Java code.

Code to serialize protocol buffers are not the only thing you want to generate. We want to generate [Retrofit](http://square.github.io/retrofit) interfaces based on RPC definitions, for example. Additionally there's no reason that you couldn't general Python, Objective-C, or even protobuf-compliant code.

The parsing of protos is easy but pruning them requires contextual knowledge of where they are stored and dependency resolution semantics is hard. Since Wire is already doing all of this work, we should expose it for others to consume and do with what they'd like.

After parsing and pruning an immutable representation of everything required for generating code should be available. We can re-use [protoparser's](http://github.com/square/protoparser) model objects since they are immutable and simple omit the messages, enums, and rpcs which have been pruned.

*(This is done on [version-two branch][branch])*

#### Alter the behavior of the Wire-specific generated Java code.

This is similar to the previous point, but involves hooking into the Wire-specific code generation rather than a completely separate generator.

For example, naming of fields and methods is a high-contention point that could easily be solved by a method which can be overridden to specify a different naming scheme.
```java
protected String toJavaName(String protoName) {
  // Note: You would want to cache the converter, obviously.
  return CaseFormat.LOWER_UNDERSCORE.converterTo(CaseFormat.LOWER_CAMEL).convert(protoName);
}
```
This would also allow for hooks to augment generation. The `JavaWriter` could be handed to empty `protected` methods before and after certain blocks of code for additional logic to be put in place. For example, additional validation code in the constructor which takes the builder.

See [Shawn's branch][hook] for a specific spike on what might be needed. Maybe we can get him to write a more fleshed out design here.

#### More advanced CLI

With all of the aforementioned changes, it should come as no surprise that a more powerful CLI is needed. By not attempting to mirror the `protoc` CLI we are unblocked to make something that's both simple, declarative, and powerful.

*more to come...*

#### Other features

 * Protos are not always on-disk and we should not require them to be. In a traditional Maven build they are often placed in src/main/protos and can be published in the artifact to Maven central or any other repository manager. We should support a way of loading them.
 * Support multiple proto directories for dependency resolution.

### TODO
 * [x] Parsing
 * [x] Pruning
 * [ ] CLI
 * [x] Emitting old Wire-specific Java code. Parsing and pruning already return an immutable model of the protos we just need to write a self-contained emitter.
 * [ ] Validation hooks for Wire Java code emitter.

 [branch]: https://github.com/square/wire/tree/version-two
 [hook]: https://github.com/square/wire/tree/shawn/emitter-hooks

**Comment by strangemonad** [comment]
is wire2 planning support for proto3?

**Comment by JakeWharton** [comment]
No. That is not part of this effort. Support for it can be tracked
separately and done as its own effort.

**Comment by bobwenx** [comment]
Are the Wire generated byte stream interoperate with other implementation? for example, on same proto definition, i am using the wire runtime on android and setup standard protobuf runtime on server side, are them wire-compatible?

**Comment by danrice-square** [comment]
Yes, the byte streams are completely compatible.

Dan
